blueprint/core/blueprint_EcmascriptEngine.h

```diff
@@ -102,68 +128,52 @@ namespace blueprint
          */
         template <typename... T>
         juce::var invoke (const juce::String& name, T... args);
 
         //==============================================================================
-        /** A public member which can be assigned a callback for delegating error handling to user code.
-         *
-         * This callback will be called primarily in the case of recoverable errors when evaluating JS code.
-         * It is possible to continue using this EcmascriptEngine instance after such an error has been
-         * raised. For example callers may wish to handle such an error and subsequently reload a modified
-         * version of a JS bundle file as part of a "hot-reload" workflow, incrementally fixing/debugging
-         * errors. ReactApplicationRoot provides such a workflow.
-         */
-        std::function<void(const juce::String& msg, const juce::String& trace)> onUncaughtError;
-
-        //==============================================================================
-        /**
-         * Pauses execution and waits for a debug client to attach and begin a debug session.
-         */
+        /** Pauses execution and waits for a debug client to attach and begin a debug session. */
         void debuggerAttach();
 
-        /**
-         * Detaches the from the current debug session/attachment.
-         */
+        /** Detaches the from the current debug session/attachment. */
         void debuggerDetach();
 
-        //==============================================================================
-        // TODO: These pushVarToDukStack/readVarFromDukStack should be private, but are
-        // made public temporarily because I hacked together a nativeMethodWrapper hook
-        // that needs it. Once we replace `nativeMethodWrapper` with the LambdaHelper
-        // stuff below, these should be made private again.
-        void pushVarToDukStack (duk_context* ctx, const juce::var& v);
-        juce::var readVarFromDukStack (duk_context* ctx, duk_idx_t idx);
-
     private:
         //==============================================================================
         void timerCallback() override;
 
-        //==============================================================================
-        // Wraps the user provided error handler and is responsible for cleaning up
-        // the EcmascriptEngine in the event of an evaluation error.
-        std::function<void()> errorHandler;
-
         //==============================================================================
         struct LambdaHelper {
-            LambdaHelper(juce::var::NativeFunction fn);
+            LambdaHelper(juce::var::NativeFunction fn, uint32_t _id);
 
             static duk_ret_t invokeFromDukContext(duk_context* ctx);
             static duk_ret_t callbackFinalizer (duk_context* ctx);
 
             juce::var::NativeFunction callback;
-            juce::Uuid                id;
+            uint32_t id;
```

Comment:
Hah. I was just musing, with the rate I'm pushing meter events through and creating temporaries I wonder how long it would take for this to overflow ...

Reply:
Yea, I thought about the same :)

My thought was basically that _most_ of the lambdas going into the release pool will be temporaries, which means that at the time the `uint32_t` wraps, we'll be fine to just start clobbering. Glad you said something though, because I'm realizing that blatantly clobbering existing entries would mean overwriting persistent functions (i.e. `registerNativeMethod` callbacks). A simple fix would be to write a `getNextId()` helper which just returns `nextId + 1` IFF the release pool is empty at that id, otherwise increment again. This, then, would only become problematic if you have so many persistent functions that you have no room left in the map for a block of temporaries (i.e. the ~30 methods of the canvas context). I think we can be reasonably sure that case would never happen!

Reply:
Fair enough, so just continue with a circular wrap and increment past any collisions when needed. Sounds like a plan. Heh. That feels like the sort of thing that warrants a comment inline so we understand what is going on in 6 months time!

Reply:
> Fair enough, so just continue with a circular wrap and increment past any collisions when needed. Sounds like a plan. Heh. That feels like the sort of thing that warrants a comment inline so we understand what is going on in 6 months time!

For sure hah. Also, realistically I should probably just benchmark this against `juce::Uuid`. If there's no significant advantage to the int then the UUID is probably easier to think about. I'm remembering that part of why I made the switch was so that I could compare `nextId` to the size of the pool, to see how many lambdas had been handed out vs how many were still allocated. But that's a debug trick, not a necessary concern in the design of this

> How do you feel about the whole `lightfunc` thing? They smell like they could be a winnder for the temporary function object's registered via `pushVarToDukStack` and `readVarFromDukStack`. I'm trying to wrap my head around whether their limitations would actually cause problems in practice here. It doesn't feel like it?

Yea, I'm with you, I just haven't taken a stab at it in here yet. Let me poke at it while I'm still in here; I think it could be worthwhile

Reply:
>I'm remembering that part of why I made the switch was so that I could compare nextId to the size of the pool, to see how many lambdas had been handed out vs how many were still allocated. But that's a debug trick, not a necessary concern in the design of this

Sneaky :-)

>Yea, I'm with you, I just haven't taken a stab at it in here yet. Let me poke at it while I'm still in here; I think it could be worthwhile

Yeah no worries. I'm happy to assist in anway with this. As far as I understand it the provide all the functionality we need for Blueprint's temporary function object's but require much less over head and no GC. It's fair to say this would require a new approach in regards to removing lightfuncs from the stack once the temporary objects are done with I assume.

Reply:
@nick-thompson, if we haven't decided on the performance improvement approach we could always pull the error handling and NativeFunction improvement out into their own PR/PRs ? I'm starting to think that might make sense as the performance stuff could be a dive down a big old rabbit hole?